File: queues/queue.py
from node import Node
# Check this page to know more about object identity:
# https://realpython.com/python-is-identity-vs-equality/
class Queue:
    def __init__(self, value):
        new_node = Node(value)
        self.first = new_node
        self.last = new_node
        self.length = 1

    def print(self):
        tmp = self.first
        while tmp is not None:
            print(tmp.value)
            tmp = tmp.next

    def enqueue(self, value):
        new_node = Node(value)
        if self.first is None and self.last is None:
            self.first = new_node
            self.last = new_node
        else:
            self.last.next = new_node
            self.last = new_node
        self.length += 1

    def dequeue(self):
        if self.first is None and self.last is None: # The stack is empty
            return None
        first = self.first
        if self.first is self.last:
            self.first = None
            self.last = None
        else:
            self.first = self.first.next
            first.next = None
        self.length -= 1
        return first

File: queues/cpp/queue.cpp
#include <iostream>
#include "queue.h"

Queue::Queue() : first(nullptr), last(nullptr), size(0) {}

Queue::Queue(int value):
first(new Node(value)),
last(nullptr),
size(1)
{
  last = first;
}

void Queue::print() const
{
  Node *tmp = first;
  while(tmp)
  {
    std::cout << tmp->value << std::endl;
    tmp = tmp->next;
  }

}

bool Queue::is_empty() const
{
  return !first && !last;
}

void Queue::enqueue(int value)
{
  Node* new_node = new Node{value};
  size++;
  if(this->is_empty())
  {
    first = new_node;
    last = new_node;
  }
  else
  {
    last->next = new_node;
    last = new_node;
  }
}

int Queue::dequeue()
{
  if(this->is_empty()) return INT_MIN;
  int value = first->value;
  Node *node = first;
  if(first == last)
  {
    first = nullptr;
    last = nullptr;
  }
  else
  {
    first = first->next;
  }
  delete node;
  size--;
  return value;
}

Queue::~Queue()
{

}
